src/clr-angular/utils/a11y/aria-live.service.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2016-2019 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+
+import { PLATFORM_ID } from '@angular/core';
+import { DOCUMENT, isPlatformBrowser } from '@angular/common';
+import { Inject, Injectable, NgZone, OnDestroy } from '@angular/core';
+import { uniqueIdFactory } from '../id-generator/id-generator.service';
+
+export enum AriaLivePoliteness {
+  off = 'off',
+  polite = 'polite',
+  assertive = 'assertive',
+}
+
+/**
+ * Time in milliseconds before inserting the content into the container
+ */
+export const ARIA_LIVE_TICK: number = 100;
+
+@Injectable()
+export class AriaLiveService implements OnDestroy {
+  private ariaLiveElement: HTMLElement;
+  private document: Document;
+  private previousTimeout: ReturnType<typeof setTimeout>;
+  private id: string = `clr-aria-live-element-${uniqueIdFactory()}`;
+
+  constructor(
+    private _ngZone: NgZone,
+    @Inject(DOCUMENT) _document: any,
+    @Inject(PLATFORM_ID) private platformId: Object
+  ) {
+    this.document = _document;
+  }
+
+  private createLiveAriaContainer(): HTMLElement {
+    const ariaLiveElement = this.document.createElement('div');
+
+    ariaLiveElement.setAttribute('id', this.getId());
+    // Use clarity screen reader class to hide the dom element
+    // and fix the scrollbar shake
+    ariaLiveElement.classList.add('clr-sr-only');
+
+    ariaLiveElement.setAttribute('aria-atomic', 'true');
+    ariaLiveElement.setAttribute('aria-live', AriaLivePoliteness.polite);
+
+    this.document.body.appendChild(ariaLiveElement);
+
+    return ariaLiveElement;
+  }
+
+  public announce(message: string | HTMLElement, politeness: AriaLivePoliteness = AriaLivePoliteness.polite): void {
```

Comment:
Is there a use case where we don't want to make an announcement? Example multiple spinner components on a page. So example instead of the component passing in the `AriaLivePoliteness .off` state it would just contain the logic to not call the `announce()` method? Wouldn't the enum still need to exist for the component so the user can disable the announcement?

Seems like if you keep the enum then the component just passes it down to the service and the service can check for off and make sure it doesn't make any unnecessary DOM manipulations. I might be misunderstanding the use case for `off`

https://github.com/vmware/clarity/pull/3934/files/ef63c2334b3a3b3be820fcfc05644073ffdf71c3#diff-4b1f682e065c8811fe25073f1cffb482R120

Reply:
Is there a use case where we don't want to make an announcement? Example multiple spinner components on a page. So example instead of the component passing in the `AriaLivePoliteness .off` state it would just contain the logic to not call the `announce()` method? Wouldn't the enum still need to exist for the component so the user can disable the announcement?

Seems like if you keep the enum then the component just passes it down to the service and the service can check for off and make sure it doesn't make any unnecessary DOM manipulations. I might be misunderstanding the use case for `off`

https://github.com/vmware/clarity/pull/3934/files/ef63c2334b3a3b3be820fcfc05644073ffdf71c3#diff-4b1f682e065c8811fe25073f1cffb482R120

Reply:
If consumer components don't want to announce anything, it simply wouldn't need to call the method from the aria-live service, right? Calling the announce method with `AriaLivePoliteness.off` itself doesn't make sense. Is it telling assistive technologies to announce the message or become silent? With `aria-life=off`, assistive technologies ignore all content change in the future.

On the side note, I haven't looked carefully into how the service is being used in consumer components. But I don't think what we have in the spinner component is correct. I will explain why... I think we need to discuss that it on Slack.

Reply:
I prefer to keep the `enum` for selecting and controlling the `politeness` of the aria. Something that I could be improved is:

```ts
  public announce(message: string | HTMLElement, politeness: AriaLivePoliteness = AriaLivePoliteness.polite): void {
    // Create aria-live container when needed.
    if (!this.ariaLiveElement && isPlatformBrowser(this.platformId)) {
      this.ariaLiveElement = this.createContainer();
    }

    // handle DOM or string input
    message = typeof message !== 'string' && isPlatformBrowser(this.platformId) ? message.textContent : message;

    // when there is no message do NOTHING!
    if (!message) {
      return;
    }

    this.ariaLiveElement.setAttribute('aria-live', politeness);
     
    if (politeness === AriaLivePoliteness.off) { return; }
   
    // ...
```
And not create new `setTimeout` and end the function there. The case with the clr-spinner is a valid one. And `off` could be used for "don't disturb" functionality. 

Going with `boolean` value to manage the state of the element is a bit hard to understand for me. I need to manage `polite`, `assertive` and `off` with `boolean` value that is `TRUE or FALSE` that is pretty ugly and may make people confused for example

```ts
ariaLiveService.announce('echo', true);
ariaLiveService.announce('echo', false);
ariaLiveService.announce('echo');
```
At first look, it looks like will announce it, the second will not, the last one .. not sure what will happen? When is `polite` and when `assertive` how to set it `off`? Just by not calling `announce` is not the answer - for example, I could have Progressbar that I don't want to announce at all but have to explain this to `clrProgressBar` and I don't have direct access to `aria-live` attribute. Maybe the progress bar is not the best example, cause it will create aria-live when there is some value to display, just empty progress bar with the value of `0` won't even create an aria-live element - here we use the last chance to start the announcing it. But still, have control in the middle of the progress to set it to off and stop it from announcing (mute it) and later on resume it without re-creating the DOM just by changing the argument - this is more efficient than removing and creating DOM.

```ts
ariaLiveService.announce('echo', 'polite');
ariaLiveService.announce('echo', 'assertive');
ariaLiveService.announce('echo', 'off');
```
Without reading the documentation I could tell what is going on. 
I went with the ENUM type but we could go with string if we really want to keep it simple. Right now we use enum that has a value of string so it's not more efficient than a simple `string` so there is no real benefit of it. But from the development point of view, it is more clear what is going on. People work better with strings than 1,0,null.

Also, don't see a huge optimization of the code by passing `boolean` and then using `if`s to handle it (branching of the code is more expensive than string values).
Our internal use of the aria-live is almost always `polite` and almost 80% of the time we don't provide an interface so it could be changed from outside. clrProgressBar, clrSpinner both components provide an option to change the `aria-live` by setting one of the attributes design do this `clrOff` `clrPolite`.

When going back on them I think that the naming is bit cryptic `clrOff` for what? but at the same time, `clrAriaLiveOff` is too long. Still, when we are providing an interface to native attribute don't see why we have to limit the values that people could set.

Reply:
I think the Enum describing the state is fine. I probably would of made the same enum as well. As for the announce method I see two options:

Use a boolean to flag whether it should toggle `assertive` or not
`ariaLiveService.announce('my message', true);`

Or use a Enum to describe the announcement type 
`ariaLiveService.announce('my message', AriaLivePoliteness.polite);`

If you really wanted to enforce that `off` to not be an option in the service but keep the enum readability then just removing it from the Enum would accomplish the same as using a boolean. I am fine with the enum API since it is just as readable and does offer a bit of minor flexibility if new announcements were needed. The components however should not call the service at all if the AriaLivePoliteness of that component is `off` as its just unnecessary. 

@bdryanovski has a good point on the component API being a bit cryptic and eventually could be realigned better in a later PR with a deprecation. Maybe the components could of had a single input using the enum values instead of two. 

Overall this API should be internal so I don't feel that it is worth holding up since both are reasonably readable.

Reply:
OK, I will keep the Enum as the second param on the `announce` method. But added a check to prevent creating or updating the container when it is off. This way NO ACTION will be done, in the same, the API will stay the same.

Reply:
The following block of code is needed in almost all consumer components just to pass the correct argument in the method. 
```
get ariaLive(): AriaLivePoliteness {
    if (isBooleanAttributeSet(this.assertive)) {
      return 'assertive';
      return AriaLivePoliteness.assertive;
    }
    if (isBooleanAttributeSet(this.off)) {
      return 'off';
      return AriaLivePoliteness.off;
    }
    return 'polite';
    return AriaLivePoliteness.polite;
  }
```
The is the main drawback I am seeing in the current pattern. If you had the argument with boolean, you could actually just call it like this:

```
ariaLiveService.announce('my message')
```
I like what Cory suggested with using a boolean to flag whether it should toggle assertive or not.

```
// it will announce politely by default
ariaLiveService.announce('my message'); 

// it will still announce but assertively now
ariaLiveService.announce('my message, true'); 
```

But even with a boolean parameter, it's not perfect. Maybe, the culprit of the issue is that we are trying to manipulate both aria-live mode as well as the text content of the element in the same method.

Reply:
Also, I just realized we let users control the aria-live mode through following input controls:
```
 @Input('clrPolite') polite: boolean = true;
 @Input('clrAssertive') assertive: boolean;
 @Input('clrOff') off: boolean;
```

Here, not only clrOff being unclear to users what it would do, the entire API will be confusing to users. I wonder why this is not a single input for this, ie: `@Input('clrAriaLive')`
In this case, we could have used that enum type like this: `@Input('clrAriaLive') ariaLive: AriaLivePoliteness`.

Reply:
Exiting on `off` is an improvement if we are going with the enum. 

> If you need to pass the third option or fourth you will have to use what? undefined and null - that will go messy really fast.

Why do you want to pass more stuff there? That's actually why things could go wrong. Methods should do only one thing and do it well. That's why I suggested that the culprit of the issue lies in trying to manipulate both aria-live mode as well as the text content of the element in the same method. Assistive technologies announce only in two different modes: `polite` or `assistive`. It's binary. There is no `off` announcement; it just mutes the future announcements. 

That being said, in the future, we could refactor and improve this method like this:

```
ariaLiveSerivce.setMode(mode: AriaLiveMode); 

ariaLiveSerivce.announce(message: string); 
```

This will allow us to set the aria-live mode of the service independently from calling the announce method, which should only change the text content. But we accept aria-live mode in separate inputs ( `@Input('clrPolite')`, `@Input('clrAssertive')`, `@Input('clrOff')` ) with boolean so we won't get much benefit from the refactoring above at this point.